feat: add-controllers - user and item

dmitry-razumov opened the pull request:
feat: add-controllers - функционал user и item

dmitry-razumov committed:
feat: add-controllers - user and item

dmitry-razumov committed:
fix: checkstyle

TarasovSam commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,34 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.CONFLICT)
+    public ErrorResponse handleValidationException(ValidationException e) {
+        return new ErrorResponse(String.format("Validation error: %s", e.getMessage()));
```
Вот здесь и в остальных местах можно дополнительно прописать стектрейс ошибки, чтобы дополнить информацию

TarasovSam commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
+    private final ItemMapper mapper;
+
+    @PostMapping
+    public ItemDto create(@RequestHeader("X-Sharer-User-Id") long userId,
```
Для эдпоинтов дополнительно можно прописать аннотацию @ResponseStatus, которая как раз и будет возвращать статус, а сам метод пусть возвращает Dto или ничего (как для удаления)
Например, для метода создания @ResponseStatus(HttpStatus.CREATED)

TarasovSam commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,22 @@
+    private long id;
+    @NotBlank(message = "name не должен быть пробелом, пустым или null")
+    private String name;
+    @NotNull(message = "description не должен быть null")
```
А описание разве может быть пустым или состоять из пробелов?

TarasovSam commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.List;
+
+@Mapper(componentModel = "spring")
+public abstract class ItemMapper {
```
@NoArgsConstructor(access = AccessLevel.PRIVATE) или [@UtilityClass](https://projectlombok.org/features/experimental/UtilityClass)
Чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса) Так это не имеет смысла ввиду того, что все в этом классе static, а значит класс не имеет состояния)
Кстати, при использовании второй, даже static использовать не потребуется)

Для всех мапперов

dmitry-razumov replied:
Заменил в mapper'ах abstract class на interface, класс в них сейчас не требуется, достаточно интерфейса
Но до конца всё равно осталось недопонимание, зачем в данном случае, когда mapper генерируется MapStruct, в случае бы необходимости использования abstract class заменять его на утилитарный статический класс
p.s. пересмотрел внимательно документацию MapStruct, везде через абстрактный класс у них реализация в таком виде...
Если бы это был не генератор MapStruct, согласен, что надежно и правильнее наверное утилитарный статический класс, если верно понимаю, чтобы можно было добавлять методы не уронив множественных наследников в первую очередь

TarasovSam replied:
Да, мой косяк (не углядел, что ты уже применил MapStruct). Хотел про него написать в следующей итерации или в следующем спринте, но ты сам его уже изучил и применил

TarasovSam commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,17 @@
+
+@Data
+@Builder
 public class Item {
```
Тут валидация не нужна?

dmitry-razumov replied:
вроде вся валидация сделана в ДТО + в сервисном слое проверки сразу на входе с контроллера дополнительные и на уровне ДАО уже сущностям вроде как и не требуется валидация

TarasovSam commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+        if (item.getName() != null) {
+            updatedItem.setName(item.getName());
+        }
+        if (item.getDescription() != null) {
```
Для имени и описания данной проверки не достаточно

TarasovSam commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
+
+    @PostMapping
+    public ItemDto create(@RequestHeader("X-Sharer-User-Id") long userId,
+                          @Valid @RequestBody ItemDto itemDto) {
```
А почему здесь не используешь @validated?)

TarasovSam commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,46 @@
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
+import java.util.List;
 
```
Что-то я у тебя нигде не увидел логирования...

dmitry-razumov replied:
добавил

TarasovSam commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,30 @@
+            message = "email не должен быть пробелом, пустым или null")
+    private String email;
+
+    public interface CreateUser {
```
Я бы интерфейсы выделил отдельно

TarasovSam commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public User update(User user) {
+        User updatedUser = storage.getById(user.getId());
+        if (user.getName() != null) {
```
Проверки для строковых полей не достаточно

TarasovSam reviewed (commented):
Дмитрий, привет!
Проверил твое задание, все довольно не плохо. Оставил несколько комментариев с тем, что можно улучшить и что можно опробовать в качестве получения опыта)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

dmitry-razumov committed:
add: stacktrace in response
add: @ResponseStatus in controllers
refact: в MapStruct mapper's interface
fix: добавлены проверки isBlank name, description в сервисном слое на входе с
контроллеров в апдейтах
add: loggin
refact: marker interfaces вынесены отдельно

TarasovSam commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -32,10 +33,18 @@ public Item update(Item item, long userId) {
         }
         if (item.getName() != null) {
-            updatedItem.setName(item.getName());
+            if (!item.getName().isBlank()) {
```
А зачем вложенные условия?

TarasovSam commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -16,31 +21,41 @@ public class UserController {
     }
 
     @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
```
Здесь все же лучше возвращать статус ОК, т.к. это значит, что операция удаления прошла успешно

TarasovSam reviewed (commented):
Дмитрий, проверил твое задание, спасибо за работу и за то, что дорабатываешь код. Оставил еще комментарии) 
Задание я принимаю, успеха в обучении и до следующего спринта!) Если будут вопросы, пиши в пачку)

dmitry-razumov committed:
fix: UserController status delete to .OK
